internal/app/server/server.go

```diff
@@ -72,10 +79,27 @@ func Run(bootstrapConfig *bootstrapv1.Bootstrap,
 	api.RegisterRouteDiscoveryServiceServer(server, gcpServer)
 	api.RegisterListenerDiscoveryServiceServer(server, gcpServer)
 
-	if mode == "serve" {
-		logger.With("address", listener.Addr()).Info(ctx, "Initializing server")
-		if err := server.Serve(listener); err != nil {
-			logger.With("err", err).Fatal(ctx, "failed to initialize server")
-		}
+	registerShutdownHandler(server, logger)
+	logger.With("address", listener.Addr()).Info(ctx, "Initializing server")
+	if err := server.Serve(listener); err != nil {
+		logger.With("err", err).Fatal(ctx, "failed to initialize server")
 	}
 }
+
+func registerShutdownHandler(server *grpc.Server, logger log.Logger) {
+	sigs := make(chan os.Signal, 1)
+	ctx := context.Background()
+	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
+	go func() {
+		sig := <-sigs
+		logger.Info(ctx, "received interrupt signal:", sig.String())
+		err := util.DoWithTimeout(ctx, func() error {
+			logger.Info(ctx, "initiating grpc graceful stop")
```

Comment:
Additionally, I think you should flush any outstanding logs, by calling `logger.Sync()`

Reply:
i see..will add that here.